beets/util/artresizer.py

```diff
@@ -83,14 +84,50 @@ def pil_resize(maxwidth, path_in, path_out=None, quality=0):
             quality = -1
 
         im.save(util.py3_path(path_out), quality=quality)
-        return path_out
+        if max_filesize > 0:
+            # If maximum filesize is set, we attempt to lower the quality of
+            # jpeg conversion by a proportional amount, up to 3 attempts
+            # First, set the maximum quality to either provided, or 95
+            if quality > 0:
+                lower_qual = quality
+            else:
+                lower_qual = 95
+            for i in range(3):
+                # 3 attempts is an abitrary choice
+                filesize = os.stat(path_out).st_size
+                log.debug(u"PIL Pass {0} : Output size: {1}B", i, filesize)
+                if filesize <= max_filesize:
+                    return path_out
+                # Calculate percentage decrease required. This is approximate,
+                # as quality may only loosely correlate with filesize.
+                frac_diff = 1 - ((filesize - max_filesize) / max_filesize)
+                assert frac_diff < 1
+                lower_qual = lower_qual * frac_diff
+                # Restrict quality dropping below 10
```

Comment:
Regarding the second question, there isn't anything on the PIL docs but I've cast it to an integer just in case. I did test this algorithm on some files a while ago but I'll try and get some example numbers to share, my aim was to get below the target file size as quickly as possible, but yes I agree we could change it later easily too. Another question that's semi-related but worth asking - do you think there'd be any preference between `maxfilesize` and `max_filesize` (just aesthetically given `maxwidth`)?

Reply:
I don't really have a clear grasp on beets' preferences here, so sorry, I don't know. I guess there is no stringent style requirement in that regard. Maybe keep `max_filesize`, since the identifier is quite long and the underscore helps readability.

Reply:
No particular institutional preference, IMO. But personally I think the more legible snake_case version is probably better. :smiley:

Reply:
I've replaced this logic with -10 quality each time - this avoids issues with files being over double the max_filesize, and attempting to compensate for the unpredictable quality/filesize relationship for JPG which will be image-dependent. Testing with a 16.4MB jpg: quality 95 = 4.1MB, 80 = 1.3MB, 70 = 877.6kB, 10 = 130.3kB